ako-infra/ingestion/avi_infra_controller.go

```diff
@@ -325,27 +318,95 @@ func fetchSEGroup(client *clients.AviClient, overrideUri ...lib.NextPage) (error
 	return nil, nil
 }
 
-func fetchVcenterServer(client *clients.AviClient) (string, error) {
-	utils.AviLog.Infof("Get vcenterserver for cloud %s", utils.CloudName)
-	uri := "/api/vcenterserver?include_name=true&cloud_ref.name=" + utils.CloudName
+func fetchVcenterServer(vCenters map[string]string, client *clients.AviClient, nextPage ...lib.NextPage) error {
+	uri := "/api/vcenterserver?include_name&cloud_ref.name=" + utils.CloudName
+	if len(nextPage) > 0 {
+		uri = nextPage[0].NextURI
+	}
 	result, err := lib.AviGetCollectionRaw(client, uri)
 	if err != nil {
-		return "", err
+		return err
 	}
 	if result.Count == 0 {
-		return "", fmt.Errorf("vcenterServer object not found")
+		return fmt.Errorf("vcenterServer object not found")
 	}
 	elems := make([]json.RawMessage, result.Count)
 	err = json.Unmarshal(result.Results, &elems)
 	if err != nil {
-		return "", err
+		return err
 	}
-	vc := models.VCenterServer{}
-	err = json.Unmarshal(elems[0], &vc)
+	for _, elem := range elems {
+		vc := models.VCenterServer{}
+		err = json.Unmarshal(elem, &vc)
+		if err != nil {
+			return err
+		}
+		vCenters[*vc.UUID] = *vc.Name
+	}
+
+	if result.Next != "" {
+		next_uri := strings.Split(result.Next, "/api/vcenterserver")
+		if len(next_uri) > 1 {
+			overrideUri := "/api/vcenterserver" + next_uri[1]
+			nextPage := lib.NextPage{NextURI: overrideUri}
+			return fetchVcenterServer(vCenters, client, nextPage)
+		}
+	}
+
+	return nil
+}
+
+func fetchClustersInVC(vcenterServerUUID string, client *clients.AviClient, nextPage ...lib.NextPage) ([]string, error) {
+	clusters := []string{}
+	uri := "/api/nsxt/clusters"
+	if len(nextPage) > 0 {
+		uri = nextPage[0].NextURI
+	}
+	var response interface{}
+	payload := map[string]string{
+		"cloud_uuid":   lib.GetCloudUUID(),
+		"vcenter_uuid": vcenterServerUUID,
+	}
+	err := lib.AviPost(client, uri, payload, &response)
 	if err != nil {
-		return "", err
+		utils.AviLog.Errorf("Faled to get NSXT Clusters, vcServer: %s, err: %s", vcenterServerUUID, err.Error())
+		return []string{}, err
 	}
-	return *vc.Name, nil
+	res, _ := response.(map[string]interface{})
+	resNSXTClusters, _ := res["resource"].(map[string]interface{})
+	resClusters, _ := resNSXTClusters["nsxt_clusters"].([]interface{})
+	for _, cluster := range resClusters {
+		cl, _ := cluster.(map[string]interface{})
+		clusters = append(clusters, cl["vc_mobj_id"].(string))
+	}
+	return clusters, nil
+}
+
+func getVCServerName(wcpClusters []string, vCenters map[string]string, client *clients.AviClient) (string, error) {
+	if len(vCenters) == 1 {
+		for _, vcName := range vCenters {
+			return vcName, nil
+		}
+	}
+	vcServerName := ""
+	clustersFoundInVC := false
+	err := fmt.Errorf("vCenterServer not found corresponding to ESX clusters %v", wcpClusters)
+	for vcUUID, vcName := range vCenters {
+		clusters, err := fetchClustersInVC(vcUUID, client)
+		if err != nil {
+			continue
+		}
+		if utils.InSlice(clusters, wcpClusters) {
```

Comment:
Can you please explain me business logic behind this? Can this functionality achieved using Set package?

Reply:
It's just checking if the elements in subslice(wcpClusters) are present in the superslice(clusters) or not. Yes, Set package can also be used as both would be ultimately using a map for comparison.